store/store_test.go

```diff
@@ -2634,270 +2634,183 @@ const customersMeValid = `
-	mockServer := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-		c.Check(r.Method, Equals, "GET")
-		c.Check(r.URL.Path, Equals, "/hello-world")
-		w.Header().Set("Content-Type", "application/hal+json")
-		w.Header().Set("X-Suggested-Currency", "EUR")
-		w.WriteHeader(http.StatusOK)
-		io.WriteString(w, MockDetailsJSON)
-		searchServerCalled = true
-	}))
-	c.Assert(mockServer, NotNil)
-	defer mockServer.Close()
-
-	purchaseServerGetCalled := false
-	purchaseServerPostCalled := false
-	mockPurchasesServer := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-		switch r.Method {
-		case "GET":
-			c.Check(r.Header.Get("Authorization"), Equals, t.expectedAuthorization(c, t.user))
-			c.Check(r.Header.Get("Accept"), Equals, jsonContentType)
-			switch r.URL.Path {
-			case ordersPath:
-				io.WriteString(w, "{}")
-			case customersMePath:
-				io.WriteString(w, customersMeValid)
-			default:
-				c.Fail()
-			}
-			purchaseServerGetCalled = true
-		case "POST":
-			c.Check(r.Header.Get("Authorization"), Equals, t.expectedAuthorization(c, t.user))
-			c.Check(r.Header.Get("Accept"), Equals, jsonContentType)
-			c.Check(r.Header.Get("Content-Type"), Equals, jsonContentType)
-			c.Check(r.URL.Path, Equals, ordersPath)
-			jsonReq, err := ioutil.ReadAll(r.Body)
-			c.Assert(err, IsNil)
-			c.Check(string(jsonReq), Equals, "{\"snap_id\":\""+helloWorldSnapID+"\",\"amount\":0.99,\"currency\":\"USD\"}")
-			w.WriteHeader(http.StatusBadRequest)
-			io.WriteString(w, `
+var buyTests = []struct {
+	SuggestedCurrency string
+	ExpectedInput     string
+	BuyStatus         int
+	BuyResponse       string
+	SnapID            string
+	Price             float64
+	Currency          string
+	ExpectedResult    *BuyResult
+	ExpectedError     string
+}{
+	{
+		// successful buying
+		SuggestedCurrency: "EUR",
+		ExpectedInput:     `{"snap_id":"` + helloWorldSnapID + `","amount":"0.99","currency":"EUR"}`,
+		BuyStatus:         http.StatusOK,
+		BuyResponse:       mockOrderResponseJSON,
+		ExpectedResult:    &BuyResult{State: "Complete"},
+	},
+	{
+		// failure due to invalid price
+		SuggestedCurrency: "USD",
+		ExpectedInput:     `{"snap_id":"` + helloWorldSnapID + `","amount":"5.99","currency":"USD"}`,
+		BuyStatus:         http.StatusBadRequest,
+		BuyResponse: `
 {
 	"error_list": [
 		{
-			"code": "xxx",
+			"code": "invalid-field",
 			"message": "invalid price specified"
 		}
 	]
-}`)
-			purchaseServerPostCalled = true
-		default:
-			c.Error("Unexpected request method: ", r.Method)
+}`,
+		Price:         5.99,
+		ExpectedError: "cannot buy snap \"hello-world\": bad request: store reported an error: invalid price specified",
+	},
+	{
+		// failure due to unknown snap ID
+		SuggestedCurrency: "USD",
+		ExpectedInput:     `{"snap_id":"invalid snap ID","amount":"0.99","currency":"EUR"}`,
+		BuyStatus:         http.StatusNotFound,
+		BuyResponse: `
+{
+	"error_list": [
+		{
+			"code": "not-found",
```

Comment:
Instead of having these strings here, can we please have two fields: errorCode and errorMessage, and cook the response in the code if errorCode is != ""?